aether-producer/producer/db.py

```diff
@@ -103,58 +131,206 @@ def make_uuid():
+            'dbname': 'offset_db_name',
+            'port': 'offset_db_port',
+            'host': 'offset_db_host',
+            'password': 'offset_db_password'
+        }
+        offset_creds = {k: SETTINGS.get(v) for k, v in offset_settings.items()}
+        global OFFSET_DB
+        offset_db_pool_size = SETTINGS.get('offset_db_pool_size', 6)
+        OFFSET_DB = PriorityDatabasePool(offset_creds, 'OffsetDB', offset_db_pool_size)
 
     @classmethod
     def update(cls, name, offset_value):
         # Update or Create if not existing
+        call = "Offset-SET"
         try:
-            session = get_session()
-            offset = session.query(cls).filter_by(
-                schema_name=name
-            ).first()
-            if offset:
-                offset.offset_value = offset_value
-                session.add(offset)
-                session.commit()
-                return offset
-            else:
-                offset = cls(
-                    schema_name=name,
-                    offset_value=offset_value
-                )
-                session.add(offset)
-                session.commit()
-                return offset
+            promise = OFFSET_DB.request_connection(0, call)  # Highest Priority
+            conn = promise.get()
+            cursor = conn.cursor()
+            query = sql.SQL(Offset.SET_STR).format(
+                schema_name=sql.Literal(name),
+                offset_value=sql.Literal(offset_value))
+            cursor.execute(query)
+            conn.commit()
+            return offset_value
         except Exception as err:
-            logger.error('Could not save offset for topic %s | %s' %
-                         (name, err))
-            return offset
+            raise err
+        finally:
+            try:
+                OFFSET_DB.release(call, conn)
+            except UnboundLocalError:
+                logger.error(f'{call} could not release a connection it never received.')
 
     @classmethod
     def get_offset(cls, name):
-        session = get_session()
-        offset = session.query(cls).filter_by(
-            schema_name=name
-        ).first()
-        return offset
+        call = "Offset-GET"
+        try:
+            promise = OFFSET_DB.request_connection(1, call)  # Lower Priority than set
+            conn = promise.get()
+            cursor = conn.cursor(cursor_factory=DictCursor)
+            query = sql.SQL(Offset.GET_STR).format(schema_name=sql.Literal(name))
+            cursor.execute(query)
+            res = [i for i in cursor]
+            if not res:
+                return None
+            return res[0][1]
+        except Exception as err:
+            raise err
+        finally:
+            try:
+                OFFSET_DB.release(call, conn)
+            except UnboundLocalError:
+                logger.error(f'{call} could not release a connection it never received.')
+
+
+class PriorityDatabasePool(object):
```

Comment:
Is there any reason for not using the psycopg connection pool classes?

http://initd.org/psycopg/docs/pool.html

Reply:
I don't believe any of those implementations allow for prioritization when requesting a connection.

Reply:
I'd read that mixing threads with greenlets could be problematic. Originally we were using SqlAlchemy's connection pooling, but it was losing track of connections and overflowing randomly for no good reason (postgres was not nearly stressed). I assumed it was for lack of explicit compatibility with greenlets so I decided to base the design on the greenlet implementation of queue which seems to be quite stable.

Reply:
in that case... :crossed_fingers: